snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,282 @@
+        except sso.ApiException as err:
+            return None, err.body
+        except sso.UnexpectedApiError as err:
+            return None, err.json_body
+
+    def upload_snap(self, name, data):
+        data['name'] = name
+        headers = {'Authorization':
+                   _macaroon_auth(self.conf, 'package_upload')}
+        response = self.post('snap-upload/', data=data, headers=headers)
+        return response
+
+    def download_snap(self, name, channel, arch, download_path,
+                      download_url, expected_sha512):
+        if self.matching_sha512(download_path, expected_sha512):
+            logger.info('Already downloaded {} at {}'.format(
+                name, download_path))
+            return
+        logger.info('Downloading {}'.format(name, download_path))
+        download = self.cpi.get(download_url)
+        with open(download_path, 'wb') as f:
+            # FIXME: Cough, we may want to buffer here (and a progress bar
+            # would be nice) -- vila 2016-04-26
+            f.write(download.content)
+        if self.matching_sha512(download_path, expected_sha512):
+            logger.info('Successfully downloaded {} at {}'.format(
+                name, download_path))
+        else:
+            raise SHAMismatchError(download_path, expected_sha512)
+
+    def matching_sha512(self, path, expected_sha512):
+        if not os.path.exists(path):
+            return False
+
+        file_sum = hashlib.sha512()
+        with open(path, 'rb') as f:
+            for file_chunk in iter(
+                    lambda: f.read(file_sum.block_size * 128), b''):
+                file_sum.update(file_chunk)
+        return expected_sha512 == file_sum.hexdigest()
+
+    # Low level helpers
+
+    def post(self, path, data, headers):
+        url = parse.urljoin(self.root_url, path)
+        response = self.session.post(url, data=data, headers=headers)
+        return response
+
+    def close(self):
+        if self.session is not None:
+            self.session.close()
+        if self.sso.session is not None:
+            self.sso.session.close()
+        if self.updown is not None:
+            self.updown.close()
+
+
+# FIXME: Refactor as a base for all clients using requests.Session() (sca,
+# updown, cpi) with (conf, root_url) to build from -- vila 2016-04-27
+class CPIClient(object):
+    """The Click Package Index knows everything about existing snaps.
+
+    https://wiki.ubuntu.com/AppStore/Interfaces/ClickPackageIndex is the
+    canonical reference.
+    """
+
+    def __init__(self, conf):
+        self.conf = conf
+        self.root_url = os.environ.get('UBUNTU_STORE_SEARCH_ROOT_URL',
+                                       UBUNTU_STORE_SEARCH_ROOT_URL)
+        if self.conf.get('package_access') is None:
+            raise InvalidCredentialsError()
+        self.session = requests.Session()
+
+    def search_package(self, snap_name, channel, arch):
```

Comment:
what about search instead of search package?
There's a long aweful discussion about how to call this things? snaps, apps, packages... We can just remove the word in here.